wsg_50_driver/src/fmf_weiss_finger.cpp

```diff
@@ -0,0 +1,444 @@
+    std::vector<double> data;
+    ros::Rate sample_rate(finger_read_rate_);
+    unsigned int sample_count = 0;
+    while(sample_count < calib_samples_ && ros::ok()) {
+        if(get_data(*cmd_mutex, data)) {
+            force_offset_ += data[0];
+            sample_count += 1;
+        }
+
+        sample_rate.sleep();
+    }
+    force_offset_ = -1*force_offset_ / sample_count; // Compute the offset
+
+    ROS_INFO("FORCE OFFSET: %f", force_offset_);
+
+    // Write the calibration result to file
+    std::ofstream cal_file;
+    cal_file.open(calibration_path_);
+    cal_file << force_offset_ << '\n';
+    cal_file.close();
+
+    return true;
+
+}
+
+/**
+ * @brief load_calibration: Load calibration, including sending params to hand if necessary
+ * @param cmd_mutex: Need to acquire this mutex before sending commands to hand
+ * @return Whether or not the calibration was successfully loaded
+*/
+bool FMFWeissFinger::load_calibration(std::mutex* cmd_mutex) {
+    if(!initialized_) {
+        ROS_ERROR("Not properly initialized, abort");
+        return false;
+    }
+
+    // Attempt to open the file
+    std::ifstream cal_file;
+    cal_file.open(calibration_path_);
+    if(!cal_file) {
+        ROS_ERROR("Could not open cal file: %s", calibration_path_.c_str());
+        return false;
+    }
+
+    // Get the calibration value
+    std::string line;
+    std::getline(cal_file, line);
+
+    // Convert to float
+    force_offset_ = std::atof(line.c_str());
+
+    cal_file.close();
+
+    return true;
+}
+
+/**
+ * @brief read_finger: Function used by read_finger_thread_
+ * @param cmd_mutex: Must acquire this mutex before communicating with hand
+*/
+void FMFWeissFinger::read_finger(std::mutex* cmd_mutex) {
+
+    ros::Rate read_rate(finger_read_rate_);
+    paused_ = false;
+
+    // Get the command to send to finger for reading
+    int cmd;
+    if(finger_id_ == 0) {
+        cmd = FINGER0_READ_CMD;
+    } else if(finger_id_ == 1) {
+        cmd = FINGER1_READ_CMD;
+    } else {
+        ROS_ERROR("Cannot get data, finger id out of range");
+        return;
+    }
+
+    // Lock for mutex, doesn't yet acquire mutex
+    std::unique_lock<std::mutex> cmd_lock(*cmd_mutex, std::defer_lock);
+    while(ros::ok() && read_finger_alive_) {
+        if(!paused_) { // Check whether we are paused
+            // Try to get the lock
+            cmd_lock.try_lock();
+            if(!cmd_lock.owns_lock()) {
+                continue;
+            }
+
+            // Create message to store result
+            wsg_50_common::WeissFingerData wfd;
+            wfd.stamp = ros::Time::now();
+
+            // Read from the hand
+            unsigned char* resp;
+            unsigned int resp_len;
+            cmd_submit(cmd, NULL, 0, true, &resp, &resp_len);
+            cmd_lock.unlock(); // Release the lock
+
+            // Push result into message
+            if(resp_len < 2) {
+                ROS_ERROR("Did not receive response from gripper");
```

Comment:
I might be wrong, but it looks like ROS here is only used for error reporting and most of this is the low level HW interface. Hopefully that means stripping out ROS should be fairly straightforward.